### PR TITLE
MoPub adapter 6.5.x implementation to support Banner Ad

### PR DESCRIPTION
**Pull request by YueVungle**
MoPub  adapter 6.5.x version is based on Vungle iOS Monetization 6.5.x release and supports Banner Ads.

### CODE REVIEW

File: Vungle/VungleRouter.m [hunk 1]
```diff
@@ -264,15 +273,24 @@ - (void)requestBannerMrecAdWithPlacementID:(NSString *)placementID delegate:(id<
         [tempDic setObject:delegate forKey:kVungleBannerDelegateKey];
         [tempDic setObject:[NSNumber numberWithInt:BannerRouterDelegateStateRequesting] forKey:kVungleBannerDelegateStateKey];
         [self.bannerDelegates addObject:tempDic];
-
+
```
Comment (crbinsf) [inline Vungle/VungleRouter.m]: Empty space

Comment (crbinsf) [reply Vungle/VungleRouter.m]: If you need the extra line here - just be sure that the change (likely caused by the standard indent) is removed.

File: Vungle/VungleRouter.m [hunk 2]
```diff
@@ -264,15 +273,24 @@ - (void)requestBannerMrecAdWithPlacementID:(NSString *)placementID delegate:(id<
...
         [tempDic setObject:[NSNumber numberWithInt:BannerRouterDelegateStateRequesting] forKey:kVungleBannerDelegateStateKey];
         [self.bannerDelegates addObject:tempDic];
-
+        
         NSError *error = nil;
-        if ([[VungleSDK sharedSDK] loadPlacementWithID:placementID error:&error]) {
-            NSLog(@"Vungle: Start to load an ad for Placement ID :%@", placementID);
+        if (CGSizeEqualToSize(size, MOPUB_MEDIUM_RECT_SIZE)) {
+            if ([[VungleSDK sharedSDK] loadPlacementWithID:placementID error:&error]) {
+                NSLog(@"Vungle: Start to load an ad for Placement ID :%@", placementID);
```
Comment (crbinsf) [inline Vungle/VungleRouter.m]: Should this be `NSLog` or `MPLogInfo`? I see that existed previously, but wanted to point this out (also, see line 288). @ashinagawa - thoughts?

File: Vungle/VungleRouter.m [hunk 3]
```diff
@@ -444,6 +463,27 @@ - (void)clearWaitingList {
     [self.waitingListDic removeAllObjects];
 }
 
+- (void)requestBannerAdFailedWithError:(NSError *)error
+                           placementID:(NSString *)placementID
+                              delegate:(id<VungleRouterDelegate>)delegate {
+    if (error) {
+        NSLog(@"Vungle: Unable to load an ad for Placement ID :%@, Error %@", placementID, error);
```
Comment (crbinsf) [inline Vungle/VungleRouter.m]: Should `NSLog` be `MPLogInfo` here?

File: Vungle/VungleRouter.m (hunk 1 above)
Comment (YueVungle) [reply Vungle/VungleRouter.m]: Got it, I will remove the change. 😀

File: Vungle/VungleRouter.m (hunk 2 above)
Comment (YueVungle) [reply Vungle/VungleRouter.m]: @crbinsf @ashinagawa Got it, I will replace all NSLogs with MPLogInfos :)

File: Vungle/VungleRouter.m (hunk 3 above)
Comment (YueVungle) [reply Vungle/VungleRouter.m]: Got it, I will replace all NSLogs with MPLogInfos :)

File: Vungle/VungleBannerCustomEvent.m [hunk 4]
```diff
@@ -90,15 +90,15 @@ - (void)vungleAdDidLoad
         self.options = options.count ? options : nil;
 
         // generate view with size
-        UIView *mrecAdView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, MOPUB_MEDIUM_RECT_SIZE.width, MOPUB_MEDIUM_RECT_SIZE.height)];
+        UIView *bannerAdView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, _bannerSize.width, _bannerSize.height)];
```
[the remark this answers is not in the document]
Comment (YueVungle) [reply Vungle/VungleBannerCustomEvent.m]: @ashinagawa Hi Aki, _bannerSize.width, _bannerSize.height and self.bannerSize.width,  self.bannerSize.height are same with each other here. And accessing instance variable is faster in read-only mode. ^_^

[the remark this answers is not in the document]
Comment (ashinagawa) [reply Vungle/VungleBannerCustomEvent.m]: I thought we had a rule that we need to access local variable through "self.xxx" outside of `init` function.

**Comment by ashinagawa** [comment]
(moved from AdMob side where I added below accidentally).

@YueVungle I noticed that some of UnitIDs in ViewController (which has not been changed in this PR tho), is set to "60d58da0b5324ae0b645670aa32104ab", "2b0d4aa5358c4190b9928601ea1f7a6a" and "de9b1104ee074542807e349f5a9ef2b5" are not belog to the same app as for other AdUnits. I think you need to prepare set of UnitIDs in the same app for all adtypes on MoPub dashboard, or this will confuse QA team.

@YueVungle I also noticed that you use your banner UnitID is linked to Vungle adppid:5c003b9a3933314cf38ff7f3 which is used for Native iOS QA Test app.
I believe you need to prepare brand-new set of appID and placement ID just for MoPub QA Test app to avoid confusion. (since you only use 1 or 2 placements under "adppid:5c003b9a3933314cf38ff7f3" but, SDK tries to auto-cache placements which are not used in MoPub dashboard, so this is very confusing for QA team.)

File: Vungle/VungleBannerCustomEvent.m (hunk 4 above)
[the remark this answers is not in the document]
Comment (crbinsf) [reply Vungle/VungleBannerCustomEvent.m]: @ashinagawa I think we've tried to use `self` when not in the `init` method. @YueVungle for consistency let's use self here.

[the remark this answers is not in the document]
Comment (YueVungle) [reply Vungle/VungleBannerCustomEvent.m]: @crbinsf @ashinagawa  Hi Clarke and Aki, thanks a lot for the reminder and telling me about the coding rule, I will update the code and use self here : )

File: Vungle/VungleRouter.m [hunk 5]
```diff
@@ -205,16 +208,22 @@ - (void)requestRewardedVideoAdWithCustomEventInfo:(NSDictionary *)info delegate:
...
 
-    // Verify if PlacementID is nil (first MREC request) or PlacementID is the same one requested
-    if (self.bannerPlacementID != nil && ![[info objectForKey:kVunglePlacementIdKey] isEqualToString:self.bannerPlacementID]) {
-
-        MPLogInfo(@"A banner ad type has been already instanciated. Multiple banner ads are not supported with Vungle iOS SDK version %@ and adapter version %@.", VungleSDKVersion, VungleAdapterVersion);
+    // Verify if PlacementID is nil (first MREC or banner request) or PlacementID is the same one requested
+    if (self.mrecPlacementID != nil && ![[info objectForKey:kVunglePlacementIdKey] isEqualToString:self.mrecPlacementID]) {
+        MPLogInfo(@"A MREC ad type has been already instantiated. Multiple MREC ads are not supported with Vungle iOS SDK version %@ and adapter version %@.", VungleSDKVersion, VungleAdapterVersion);
+        [delegate vungleAdDidFailToLoad:nil];
+    } else if (self.bannerPlacementID != nil && ![[info objectForKey:kVunglePlacementIdKey] isEqualToString:self.bannerPlacementID]) {
```
Comment (ashinagawa) [inline Vungle/VungleRouter.m]: As I wrote in AdMob side, we might want to set `nil` for `bannerPlacementID` and `mrecPlacementID` after finishing displaying banner/mrec ad, so `router` class can allow to load for next banner/mrec ad with "different" placement ID.

Comment (YueVungle) [reply Vungle/VungleRouter.m]: @ashinagawa Hi Aki, according to your comments, I set nil for bannerPlacementID and mrecPlacementID after finishing displaying banner/mrec ad for both MoPub and AdMob Adapter, please review the changes in these two PRs. Thanks a lot. :)

Comment (ashinagawa) [reply Vungle/VungleRouter.m]: @YueVungle Please verify "refresh" is working correctly after your update.

Comment (YueVungle) [reply Vungle/VungleRouter.m]: @ashinagawa Hi Aki, I have verified that "refresh" works correctly after the update here. :)
It seems like this PR has passed the code review, may I close this RR :)

**Pull request closed without merging**